Mathlib/Algebra/Lie/Weights/Chain.lean

```diff
@@ -0,0 +1,130 @@
+`α`-chain through `χ`.
+
+We provide basic definitions and results to support `α`-chain techniques in this file.
+
+## Main definitions / results
+
+ * `LieModule.weightSpaceChain`: given weights `χ₁`, `χ₂` together with integers `p` and `q`, this
+   is the sum of the weight spaces `k • χ₁ + χ₂` for `p < k < q`.
+ * `LieModule.trace_toEndomorphism_weightSpaceChain_eq_zero`: given a root `α` relative to a Cartan
+   subalgebra `H`, there is a natural ideal `(rootSpaceProductNegSelf α).range` in `H`. This lemma
+   states that this ideal acts by trace-zero endomorphisms on the sum of root spaces of any
+   `α`-chain, provided the weight spaces at the endpoints are both trivial.
+
+-/
+
+open Set
+
+variable {R L : Type*} [CommRing R] [LieRing L] [LieAlgebra R L]
+  (M : Type*) [AddCommGroup M] [Module R M] [LieRingModule L M] [LieModule R L M]
+
+namespace LieModule
+
+section IsNilpotent
+
+variable [LieAlgebra.IsNilpotent R L] (χ₁ χ₂ : L → R) (p q : ℤ)
+
+/-- Given two (potential) weights `χ₁` and `χ₂` together with integers `p` and `q`, it is often
+useful to study the sum of weight spaces associated to the family of weights `k • χ₁ + χ₂` for
+`p < k < q`. -/
+def weightSpaceChain : LieSubmodule R L M :=
+  ⨆ k ∈ Ioo p q, weightSpace M (k • χ₁ + χ₂)
+
+lemma weightSpaceChain_def :
+    weightSpaceChain M χ₁ χ₂ p q = ⨆ k ∈ Ioo p q, weightSpace M (k • χ₁ + χ₂) :=
+  rfl
+
+lemma weightSpaceChain_def' :
+    weightSpaceChain M χ₁ χ₂ p q = ⨆ k ∈ Finset.Ioo p q, weightSpace M (k • χ₁ + χ₂) := by
+  have : ∀ (k : ℤ), k ∈ Ioo p q ↔ k ∈ Finset.Ioo p q := by simp
+  simp_rw [weightSpaceChain_def, this]
+
+@[simp]
+lemma weightSpaceChain_neg :
+    weightSpaceChain M (-χ₁) χ₂ (-q) (-p) = weightSpaceChain M χ₁ χ₂ p q := by
+  let e : ℤ ≃ ℤ := neg_involutive.toPerm
+  simp_rw [weightSpaceChain, ← e.biSup_comp (Ioo p q)]
+  simp [-mem_Ioo, neg_mem_Ioo_iff]
+
+end IsNilpotent
+
+section IsNilpotentSubalgebra
+
+open LieAlgebra
+
+variable {H : LieSubalgebra R L} (α χ : H → R) (p q : ℤ)
+
+lemma lie_mem_weightSpaceChain_of_weightSpace_eq_bot_right [LieAlgebra.IsNilpotent R H]
+    (hq : weightSpace M (q • α + χ) = ⊥)
+    {x : L} (hx : x ∈ rootSpace H α)
+    {y : M} (hy : y ∈ weightSpaceChain M α χ p q) :
+    ⁅x, y⁆ ∈ weightSpaceChain M α χ p q := by
+  rw [weightSpaceChain, iSup_subtype'] at hy
+  induction' hy using LieSubmodule.iSup_induction' with k z hz z₁ z₂ _ _ hz₁ hz₂
+  · obtain ⟨k, hk⟩ := k
+    suffices weightSpace M ((k + 1) • α + χ) ≤ weightSpaceChain M α χ p q by
+      apply this
+      simpa using (rootSpaceWeightSpaceProduct R L H M α (k • α + χ) ((k + 1) • α + χ)
+        (by rw [add_smul]; abel) (⟨x, hx⟩ ⊗ₜ ⟨z, hz⟩)).property
+    rw [weightSpaceChain]
+    rcases eq_or_ne (k + 1) q with rfl | hk'; · simp only [hq, bot_le]
+    replace hk' : k + 1 ∈ Ioo p q := ⟨by linarith [hk.1], lt_of_le_of_ne hk.2 hk'⟩
+    exact le_biSup (fun k ↦ weightSpace M (k • α + χ)) hk'
```

Comment:
this should probably be its own lemma

Reply:
I don't think it should since this is a weaker version of what we're proving.

However your remark has inspired me to add the result `LieModule.weightSpace_le_weightSpaceChain` and using that I can emphasise my point:
```lean
example [LieAlgebra.IsNilpotent R H] {k : ℤ} (hk : k ∈ Ioo p q)
    (hq : weightSpace M (q • α + χ) = ⊥)
    {x : L} (hx : x ∈ rootSpace H α)
    {y : M} (hy : y ∈ weightSpace M (k • α + χ)) :
    ⁅x, y⁆ ∈ weightSpaceChain M α χ p q :=
  lie_mem_weightSpaceChain_of_weightSpace_eq_bot_right M α χ p q hq hx <|
    weightSpace_le_weightSpaceChain M α χ p q hk hy
```

Reply:
Sorry I don't think I was clear, I did actually mean `weightSpace_le_weightSpaceChain` - I know it's absolutely clear how `weightSpaceChain` is defined to anyone using it, but I still think we should have the API lemmas instead of forcing people to look though the `biSup`.

Reply:
OK great, so all is well now?